src/v/cluster/metadata_cache.cc

```diff
@@ -248,4 +272,69 @@ metadata_cache::get_default_shadow_indexing_mode() const {
     }
     return m;
 }
+
+ss::future<> metadata_cache::broadcast_disk_space_health(
+  const storage::disk_space_alert state) {
+    return container().invoke_on_others(
+      [state](metadata_cache& mc) { mc.set_disk_space_health(state); });
+}
+
+void metadata_cache::set_disk_space_health(const storage::disk_space_alert s) {
+    _disk_space_health = s;
+}
+
+bool metadata_cache::should_reject_writes() const {
+    return _disk_space_health == storage::disk_space_alert::degraded;
+}
+
+ss::future<>
+metadata_cache::update_disk_health_cache(const cluster_health_report& chr) {
+    storage::disk_space_alert disk_health{};
+
+    for (auto& r : chr.node_reports) {
+        vlog(
+          clusterlog.trace,
+          "{}: node {} {}",
+          __func__,
+          r.id,
+          r.local_state.storage_space_alert);
+        disk_health = std::max(r.local_state.storage_space_alert, disk_health);
+    }
+    vlog(clusterlog.debug, "{} -> {}", __func__, disk_health);
+    set_disk_space_health(disk_health);
+    co_await broadcast_disk_space_health(disk_health);
+}
+
+// Handler for refresh_shard's update timer
+void metadata_cache::update_disk_health() {
+    if (_refresh_gate.is_closed()) {
+        return;
+    }
+    ssx::spawn_with_gate(_refresh_gate, [this]() -> ss::future<> {
+        // Ensure that this node's cluster health data is not too stale.
+        co_await _health_monitor.local()
+          .get_cluster_health(
+            {.node_report_filter
+             = {.include_partitions = include_partitions_info::no}},
+            force_refresh::no,
+            model::time_from_now(health_monitor_frontend::default_timeout))
+          .then([this](result<cluster_health_report> res) -> ss::future<> {
```

Comment:
it's a bit unclear why polling is necessary. would it not work to refresh the health monitor state if it is found to be out-dated and need of refresh?

it would be nice to keep the metadata cache as a place where data is pushed into it.

but maybe there is a design aspect here I'm missing.

Reply:
Health monitor only ticks on ring0 leader. Are you suggesting changing health monitor to always tick on all nodes? Also, are you suggesting we make cross-shard requests in the produce handler fast path instead of using a background timer here?

Reply:
> Health monitor only ticks on ring0 leader. Are you suggesting changing health monitor to always tick on all nodes? 

Not quite, but close. When ring0 leader tick discovers a change it can propagate the information all all cores so that it is available via the fast path. I think that is very close to what you are doing now, but instead of pulling the state up into the metadata cache with another poller, keep it all in the health manager and push the updates to the other cores. Still maybe there is something I'm missing here.

Reply:
Ring 0 leader is a node, not a shard, so having it propagate to all cores only gives you fresh cluster health reports on a single node in the cluster.

Reply:
We discussed this a bit; I'm going to try and add a new per-node refresh timer to health monitor front end. If it is not too messy, we'll go with that and eliminate the changes to metadata cache.